dnc/util.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""DNC util ops and modules."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+
+
+def batch_invert_permutation(permutations):
+  """Returns batched `tf.invert_permutation` for every row in `permutations`."""
+  with tf.name_scope('batch_invert_permutation', values=[permutations]):
+    # unpacked = tf.unstack(permutations)
+    # inverses = [tf.invert_permutation(permutation) for permutation in unpacked]
+    # return tf.stack(inverses)
+
+    # fix unknown shape issue when using tf.unstack
+    perm = tf.cast(permutations, tf.float32)
+    dim = int(perm.get_shape()[-1])
+    size = tf.cast(tf.shape(perm)[0], tf.float32)
+    delta = tf.cast(tf.shape(perm)[-1], tf.float32)
+    rg = tf.range(0, size*delta, delta, dtype=tf.float32)
+    rg = tf.reshape(rg, [-1, 1])
```

Comment:
nit: I think tf.expand_dims(rg, 1) is a bit clearer here, as you're specifically trying to expand out a dimension.